boot/bootutil/src/loader.c

```diff
@@ -1916,3 +1926,216 @@ split_go(int loader_slot, int split_slot, void **entry)
     free(sectors);
     return rc;
 }
+
+#else /* !MCUBOOT_DIRECT_XIP */
+
+/**
+ * Iterates over all slots and determines which contain a firmware image.
+ *
+ * @param state          Boot loader status information.
+ * @param slot_usage     A pointer to a zero-initialized array, which aim is to
+ *                       carry information about slots that contain an image.
+ *                       After return the corresponding array elements are set
+ *                       to a non-zero value if the given slots are in use
+ *                       (contain a firmware image).
+ * @param slot_cnt       The number of slots, which can contain firmware images.
+ *
+ * @return               The number of found images.
+ */
+static uint32_t
+boot_get_slot_usage(struct boot_loader_state *state, uint32_t *slot_usage,
+                    uint32_t slot_cnt)
+{
+    struct image_header *hdr = NULL;
+    uint32_t image_cnt = 0;
+    uint32_t slot;
+
+    for (slot = 0; slot < slot_cnt; slot++) {
+        hdr = boot_img_hdr(state, slot);
+
+        if (boot_is_header_valid(hdr, BOOT_IMG_AREA(state, slot))) {
+            slot_usage[slot] = 1;
+            image_cnt++;
+            BOOT_LOG_IMAGE_INFO(slot, hdr);
+        } else {
+            BOOT_LOG_INF("%s slot: Image not found", (slot == BOOT_PRIMARY_SLOT)
+                         ? "Primary" : "Secondary");
+        }
+    }
+
+    return image_cnt;
+}
+
+/**
+ * Prepares the booting process. This function choose the newer image in flash
+ * as appropriate, and tells you what address to boot from.
+ *
+ * @param state                 Boot loader status information.
+ * @param rsp                   On success, indicates how booting should occur.
+ *
+ * @return                      0 on success; nonzero on failure.
+ */
+int
```

Comment:
I did not compare to the current `context_boot_go` but this looks fairly similar. Maybe deserves a medal for the copy/pasta Olympics! :-)

Reply:
It is fairly different, but there is a lot in common. It might be worth trying to factor this function into what is common and what is unique to each.

Reply:
I think in its current state it would be better to keep them separate, I'm planning to add some extra code to the direct-xip code in the near future to support another upgrade strategy (RAM_LOAD). It would be possible (maybe a little forced) to merge the 2 context_boot_go() functions, but it would make the first even longer and would require more ifdefs.
Common parts:
- 3 local variables
- a short 'for' loop to open the image areas
- **the ends of the functions are the most similar**:
  - security counter update (HW rollback protection): can be combined with a few extra ifdefs
  - MCUBOOT_MEASURED_BOOT and MCUBOOT_DATA_SHARING: can be combined if parameterized (selected_slot)
  - Setting the 'rsp' structure: can be combined if parameterized (selected_slot +slected_image_header?)
- a short 'for' loop to close the image areas

Reply:
>  encrypted images won't work

Our format of encrypted images are unlikely to work, but there are more and more SoCs coming out with the ability to execute encrypted code. Usually the key management is a bit more complex, because the key never is visible to the CPU. Regardless, that is the topic of a different change than this one.

Reply:
Are there any more thoughts on this topic? What is your opinion, can I leave it this way or shall I try to merge the 2 functions?